File: go/0205_isomorphic_strings.go
package main

// time O(n)
// space O(n)

func isIsomorphic(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	hashST := make(map[byte]byte)
	hashTS := make(map[byte]byte)

	for i := 0; i < len(s); i++ {
		sChar, tChar := s[i], t[i]
		if sMappedChar, exists := hashST[sChar]; exists && sMappedChar != tChar {
			return false
		}
		if tMappedChar, exists := hashTS[tChar]; exists && tMappedChar != sChar {
			return false
		}
		hashST[sChar] = tChar
		hashTS[tChar] = sChar
	}

	return true
}

File: csharp/0205-isomorphic-strings.cs
public partial class Solution {
    public bool IsIsomorphic(string s, string t) {
        if (s.Length != t.Length) {
            return false;
        }

        var hashSt = new Dictionary<char, char>();
        var hashTs = new Dictionary<char, char>();

        for (var i = 0; i < s.Length; i++) {
            var sChar = s[i];
            var tChar = t[i];
            if (hashSt.ContainsKey(sChar) && hashSt[sChar] != tChar) {
                return false;
            }
            if (hashTs.ContainsKey(tChar) && hashTs[tChar] != sChar) {
                return false;
            }
            hashSt[sChar] = tChar;
            hashTs[tChar] = sChar;
        }

        return true;
    }
}